src/GafferScene/Rename.cpp

```diff
@@ -0,0 +1,760 @@
+	InternedStringVectorDataPtr result = new InternedStringVectorData( inPlug()->childNamesPlug()->getValue()->readable() );
+	for( auto &name : result->writable() )
+	{
+		auto it = nameMap.find( name );
+		if( it != nameMap.end() )
+		{
+			name = it->outputName;
+		}
+	}
+
+	return result;
+}
+
+void Rename::hashSet( const IECore::InternedString &setName, const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const
+{
+	// This hash is a brute force implementation that matches `computeSet()` closely.
+	// Things in it's favour :
+	//
+	// - If the intersection of the set and the filter is small, it visits few locations.
+	// - It never visits more locations then needed.
+	// - It is accurate, and doesn't lead to more computes than necessary.
+	//
+	// Other possibilities include :
+	//
+	// 1. Computing a one-off "hash of all the renamed locations" on another plug, so
+	//    that the hash for any set is just the input hash plus the hash of renamed
+	//    locations. This might do more work than necessary if the sets required don't
+	//    intersect the filter. But in the case of many sets covering large portions of
+	//    of the scene it might be a win?
+	//
+	// 2. Doing a "poor man's hash" where we hash the filter for the whole scene (cheap
+	//    for PathFilter and SetFilter) and use `dirtyCount()` and `Context::hash()` to
+	//    represent the values delivered by `namePlug()`. This would be fast, but would
+	//    potentially expose us to a lot of unnecessary recomputes.
+	//
+	// We currently anticipate the number of locations hit by a single Rename node to
+	// be relatively small, so hopefully the current strategy is reasonable.
+
+	const MurmurHash inputSetHash = inPlug()->setPlug()->hash();
+	ConstPathMatcherDataPtr inputSetData = inPlug()->setPlug()->getValue( &inputSetHash );
+
+	using ThreadLocalHash = tbb::enumerable_thread_specific<IECore::MurmurHash>;
+
+	struct LocationProcessor
+	{
+
+		LocationProcessor( const Rename *rename, const PathMatcher &inputSet, ThreadLocalHash &hash )
+			:	m_rename( rename ), m_parent( nullptr ), m_inputSet( inputSet ), m_hash( hash )
+		{
+		}
+
+		LocationProcessor( const LocationProcessor &parent )
+			:	m_rename( parent.m_rename ), m_parent( &parent ), m_inputSet( parent.m_inputSet ), m_hash( parent.m_hash )
+		{
+		}
+
+		bool operator()( const ScenePlug *scene, const ScenePlug::ScenePath &path )
+		{
+			const int setMatch = m_inputSet.match( path );
+			if( !( setMatch & ( PathMatcher::ExactMatch | PathMatcher::DescendantMatch ) ) )
+			{
+				return false;
+			}
+
+			const int filterMatch = m_rename->filterValue( Context::current() );
+			if( m_parent && (filterMatch & PathMatcher::ExactMatch) )
+			{
+				const NameMapData::Map &nameMap = m_parent->m_nameMap->map;
+				auto it = nameMap.find( path.back() );
+				if( it != nameMap.end() )
+				{
+					MurmurHash renameHash;
+					renameHash.append( path );
+					renameHash.append( it->outputName );
```

Comment:
I gave that a go, and although timings are a bit noisy I think it does give a measurable improvement in the order of around 4% on the Moana rename. Since we looking at such small margins, I'm inclined to leave it in favour of code clarity and real-world-lead profiling if necessary later.

I did actually spend a fair amount of time on a scheme of my own for speeding this up and/or reducing memory usage - only storing the shortest renamed prefix in `__inputPath`. But although it was showing a bit of promise, I opted to keep things simple for an initial release. Until this has seen real-world usage, I'm more concerned that there'll be an error in the correctness of my implementation that I am about a bit of performance that might be left on the floor.